### PR TITLE
ggml : fuse soft_max sweeps into fewer passes

### PR DESCRIPTION
**Pull request by cschanaj**
## Overview
  
Lossless speedup of the CPU `soft_max` op (`ggml_compute_forward_soft_max_f32`) that reduces its memory traffic. `GGML_OP_SOFT_MAX` runs in the attention-score softmax (when flash
attention is off), MoE expert routing (every MoE layer and token), and the sampler's logits softmax (top-k/top-p; greedy argmax skips it).

- Folds copy + scale + mask into a single vectorized write to `dst` (the same `GGML_F32_VEC` macros as `ggml_vec_scale_f32` — hand-SIMD on AVX-512/AVX2/SSE2/NEON/SVE/RVV)
- Runs the `exp` pass in place (`y == x` is safe: each output lane depends only on the scalar `max`), and uses no per-thread workspace (neither the buffer nor the `ggml_graph_plan` reservation). 
- The f16-mask path folds copy + scale + the f16→f32 mask-add into a single scalar loop (f16→f32 blocks SIMD for the mask-add); only T5/T5Encoder self-attention reaches it. 

The graph is unchanged — a single `SOFT_MAX` node, kernel-internal. Output matches the reference within `test-backend-ops` tolerance. Modest performance gain, bigger rows benefit more than short ones.
 
## Additional information
  
**Correctness:** all `SOFT_MAX` cases pass in `test-backend-ops` on CPU (212/212: mask/no-mask, f16 and f32 mask, ALiBi, sinks, inplace).
  
  **Performance** (`test-backend-ops perf -o SOFT_MAX -b CPU`), AMD Ryzen 9 5950X (Zen3/AVX2). Values are µs/run; **Δ is from the median of 3 clean back-to-back runs (p1, p3, p5)** — the  runs that were clean for both master and branch. Two earlier runs hit intermittent memory contention (p2 on branch, p4 on master) and were excluded; the median is robust to such outliers:
  
  | shape (ne) | mask | master (p1/p3/p5) | branch (p1/p3/p5) | Δ (median) |
  |---|---|---|---|---|
  | `[524288,16,1,1]` (batched vocab softmax) | — | 3912 / 3850 / 3793 | 851 / 797 / 862 | **−78%** |
  | `[524288,1,1,1]` (per-token vocab softmax) | — | 413 / 409 / 411 | 371 / 386 / 388 | −6% |
  | `[262144,1,1,1]` (per-token vocab softmax) | — | 209 / 211 / 210 | 185 / 193 / 195 | −8% |
  | `[1024,1024,10,1]` | — | 3045 / 2822 / 2764 | 2734 / 2786 / 2373 | −3% |
  | `[77,4096,5,1]` | — | 340 / 377 / 327 | 251 / 189 / 184 | −44% |
  | `[64,64,20,1]` | — | 31.0 / 22.9 / 21.3 | 22.7 / 16.7 / 16.8 | −27% |
  | `[512,1,16,1]` (T5 decode) | f16 | 10.3 / 10.0 / 10.2 | 11.3 / 10.8 / 11.0 | +0.8 µs (+7%) |
  | `[512,512,16,1]` (T5 prefill) | f16 | 903 / 670 / 767 | 592 / 730 / 746 | −5% |

The big-row win is removing the per-thread `wp` workspace (~2 MiB/thread at `ne00=524288`, ×16 ≈ 32 MiB of cache-thrashing the in-`dst` write avoids). The two f16 rows are the only shapes that exercise the f16-mask path (T5); decode is +0.6 µs on a µs-scale op, prefill sits at the L3 boundary and is noisy for both master and branch (median −19%). `[4096,4096,5,1]` is flat (DRAM-saturated). The two f16 shapes are committed as perf cases in `make_test_cases_perf` for reproducibility.
  
**Reproduce:**
```bash
cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DGGML_NATIVE=ON -DGGML_CUDA=OFF
cmake --build build --target test-backend-ops -j
./build/bin/test-backend-ops -o SOFT_MAX -b CPU          # correctness: expect 212/212
./build/bin/test-backend-ops perf -o SOFT_MAX -b CPU      # perf, us/run, master vs branch
```

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES - used to analyze the softmax kernel, implement and vectorize the change, testing and running, summarizing benchmark results.

### CODE REVIEW

**Comment by taronaeo** [comment]
@cschanaj Can you have a look at the bot's review and let us know if any of it is imprecise? https://github.com/ggml-org/llama.cpp/pull/26468#issuecomment-5168438466 :)

**Comment by cschanaj** [comment]
done.  I updated the PR description and added tests for the f16 path. re-ran perf to produces more benchmark results.

**Comment by cschanaj** [comment]
seemingly something was wrong with the p2 perf data so I ran more tests and updated the description. I'm suspecting the new implementation might be more sensitive to cpu busyness (i.e. performance regress back into master's baseline when there are contention, which makes sense). this PR should be ready for review now.  thanks in advance!

**Comment by cschanaj** [comment]
@taronaeo would you be able to take a second look or let me know if anything else is needed from my side?
